### ISSUE TITLE
Consider always discarding __device__ lambdas' results

### ISSUE THREAD

**Issue by jaredhoberock** [thrust/thrust#779]
We can't rely on the result of `std::result_of` when applied to a `__device__` lambda.

If we have the means to detect their presence, then one option is to entirely ban them from `bulk_invoke` and friends. A more relaxed policy might be to always interpret their result to be `void`. That would still allow them to be used, but it would not be possible to return results from them.

**Comment by brycelelbach** [comment]
I don't think there's a way to detect that it's a lambda.

I'm not sure I understand why result_of doesn't work in this instance?

Would decltype work?

**Comment by jaredhoberock** [comment]
The issue is that `__device__` lambdas' *always* return `int`! Even if there is no `return` used inside at all. Ugh! That means utilities like `result_of` or `invoke_result` or `decltype` to can't reliably determine what `__device__` lambdas return. :-(

Fortunately, we are able to [distinguish `__device__` lambdas from other types](https://github.com/agency-library/agency/blob/master/agency/detail/type_traits.hpp#L178).

Agency uses this inside its [implementation of `result_of`](https://github.com/agency-library/agency/blob/master/agency/detail/type_traits.hpp#L192) to treat all `__device__` lambdas as if they return `void`.

**Comment by allisonvacanti** [comment]
I'm going to start using this issue to track the various "`__device__` lambda aren't working" issues.

The issues Jared mentioned above are documented [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#extended-lambda-restrictions) in the current toolkit documentation.

Until we have a better system in place, users should stick with `__host__ __device__` lambda or explicit `__device__` functor objects instead.

**Comment by jrhemstad** [comment]
Just for completeness, the exact restriction is G.6.2p13:

>As described above, the CUDA compiler replaces a _device_ extended lambda defined in a host function with a placeholder type defined in namespace scope. This placeholder type does not define a operator() function equivalent to the original lambda declaration. *An attempt to determine the return type or parameter types of the operator() function may therefore work incorrectly in host code, as the code processed by the host compiler will be semantically different than the input code processed by the CUDA compiler.*

I've stopped using `__device__` lambdas all together with Thrust after I learned about this restriction. Too great a chance of silent or non-obvious failures.

**Comment by harrism** [comment]
@jaredhoberock said "device lambdas always return int". That can't be true, I have used device lambdas that return non-int before. I guess what Jared means is that `result_of` on a device lambda always evaluates to `int`?

**Comment by allisonvacanti** [comment]
Correct. The device lambda object is replaced with a placeholder object in host code, and that placeholder's API does not accurately represent the device lambda's actual call signature.

**Comment by jrhemstad** [comment]
> I guess what Jared means is that `result_of` on a device lambda always evaluates to `int`

https://godbolt.org/z/6ob7q5

Not necessarily, no. It's moreso that the return type of the callable cannot be relied upon to be correct.

**Comment by harrism** [comment]
I just wanted reassurance that if I call a device lambda that returns float from a kernel, it doesn't return int.

**Comment by pauleonix** [comment]
> I'm going to start using this issue to track the various "`__device__` lambda aren't working" issues.

https://github.com/NVIDIA/thrust/issues/1650#issuecomment-1339021709 has a reproducer for an issue with both `__device__` and `__host__ __device__` lambdas causing `thrust::transform_output_iterator< ... >::operator=` to be deleted. Further information in https://github.com/NVIDIA/thrust/issues/1650#issuecomment-1339668169
